### PR TITLE
MAINT: Cleanup `ma.array.__str__`

### PR DESCRIPTION
**Pull request by eric-wieser**
Remove hack from gh-7659 for gh-7493

### CODE REVIEW

**Comment by eric-wieser** [comment]
Clumsy mistake first time around on that last commit. All passing now.

**Comment by eric-wieser** [comment]
Actually, this is a bugfix. Before (1.13):
```
>>> x = np.ma.array((1, '2006'), dtype=[('a', int), ('b', 'M8[Y]')])
>>> x
masked_array(data = (1, '2006'),
             mask = (False, False),
       fill_value = (999999, 'NaT'),
            dtype = [('a', '<i4'), ('b', '<M8[Y]')])

>>> x['a'] = np.ma.masked
>>> x
masked_array(data = (--, datetime.date(2006, 1, 1)),
             mask = ( True, False),
       fill_value = (999999, 'NaT'),
            dtype = [('a', '<i4'), ('b', '<M8[Y]')])
```

After

```
In [1]: x = np.ma.array((1, '2006'), dtype=[('a', int), ('b', 'M8[Y]')])

In [2]: x
Out[2]:
masked_array(data = (1, datetime.date(2006, 1, 1)),
             mask = (False, False),
       fill_value = (999999, 'NaT'),
            dtype = [('a', '<i4'), ('b', '<M8[Y]')])

In [3]: x['a'] = np.ma.masked

In [4]: x
Out[4]:
masked_array(data = (--, datetime.date(2006, 1, 1)),
             mask = ( True, False),
       fill_value = (999999, 'NaT'),
            dtype = [('a', '<i4'), ('b', '<M8[Y]')])
```

Although part of that is probably the 0d scalar stuff, and that'll likely change again with #9332

**Review by mhvk — APPROVED** [review]
Look all OK.

**Comment by charris** [comment]
@mhvk Feel free to merge ...

**Comment by ahaldane** [comment]
Give me a few minutes to finish looking at this.. I started yesterday but there are updates since then.

File: numpy/ma/core.py [hunk 1]
```diff
@@ -2418,12 +2418,13 @@ def _recursive_printoption(result, mask, printopt):
 
     """
     names = result.dtype.names
-    for name in names:
-        (curdata, curmask) = (result[name], mask[name])
-        if curdata.dtype.names:
+    if names:
```
Comment (ahaldane) [inline numpy/ma/core.py]: technically not necessary since this function is always called with names (or else the old code would fail). Fine to leave it.

Comment (ahaldane) [reply numpy/ma/core.py]: OK, I get that this was intentional based on reorganization below.

**Review by ahaldane — APPROVED** [review]
All right, looks good.

**Comment by eric-wieser** [comment]
Two approvals for a MAINT sounds like a good enough excuse to self-merge to me.

**Pull request merged**